go/lang/security/audit/database/string-formatted-query.yaml

```diff
@@ -53,117 +53,48 @@ rules:
-        $QUERY = $FXN(..., $QUERY, ...)
-        ...
-        $OBJ.Query($QUERY, ...)
-    - pattern: |
-        $QUERY = "..."
-        ...
-        $QUERY = $FXN(..., $QUERY, ...)
-        ...
-        $OBJ.ExecContext($CTX, $QUERY, ...)
-    - pattern: |
-        $QUERY = "..."
-        ...
-        $QUERY = $FXN(..., $QUERY, ...)
-        ...
-        $OBJ.QueryContext($CTX, $QUERY, ...)
-    - pattern: |
-        $QUERY = "..."
-        ...
-        $QUERY = $FXN(..., $QUERY, ...)
-        ...
-        $OBJ.QueryRow($QUERY)
-    - pattern: |
-        $QUERY = "..."
-        ...
-        $QUERY = $FXN(..., $QUERY, ...)
-        ...
-        $OBJ.QueryRow($CTX, $QUERY)
-    - pattern: |
-        $QUERY = "..."
-        ...
-        $QUERY = $FXN(..., $QUERY, ...)
-        ...
-        $OBJ.QueryRowContext($CTX, $QUERY, ...)
-    - pattern: |
-        $QUERY = "..."
-        ...
-        $OTHER = $FXN(..., $QUERY, ...)
-        ...
-        $OBJ.Exec($OTHER, ...)
-    - pattern: |
-        $QUERY = "..."
-        ...
-        $OTHER = $FXN(..., $QUERY, ...)
-        ...
-        $OBJ.Query($OTHER, ...)
-    - pattern: |
-        $QUERY = "..."
-        ...
-        $OTHER = $FXN(..., $QUERY, ...)
-        ...
-        $OBJ.ExecContext($CTX, $OTHER, ...)
-    - pattern: |
-        $QUERY = "..."
-        ...
-        $OTHER = $FXN(..., $QUERY, ...)
-        ...
-        $OBJ.QueryContext($CTX, $OTHER, ...)
-    - pattern: |
-        $QUERY = "..."
-        ...
-        $OTHER = $FXN(..., $QUERY, ...)
-        ...
-        $OBJ.QueryRow($OTHER)
-    - pattern: |
-        $QUERY = "..."
-        ...
-        $OTHER = $FXN(..., $QUERY, ...)
-        ...
-        $OBJ.QueryRow($CTX, $OTHER)
-    - pattern: |
-        $QUERY = "..."
-        ...
-        $OTHER = $FXN(..., $QUERY, ...)
-        ...
-        $OBJ.QueryRowContext($CTX, $OTHER, ...)
-    - pattern: |
-        $QUERY = $X + ...
-        ...
-        $OBJ.Exec($QUERY, ...)
-    - pattern: |
-        $QUERY = $X + ...
-        ...
-        $OBJ.Query($QUERY, ...)
-    - pattern: |
-        $QUERY = $X + ...
-        ...
-        $OBJ.ExecContext($CTX, $QUERY, ...)
-    - pattern: |
-        $QUERY = $X + ...
-        ...
-        $OBJ.QueryContext($CTX, $QUERY, ...)
-    - pattern: |
-        $QUERY = $X + ...
-        ...
-        $OBJ.QueryRow($QUERY)
-    - pattern: |
-        $QUERY = $X + ...
-        ...
-        $OBJ.QueryRow($CTX, $QUERY)
-    - pattern: |
-        $QUERY = $X + ...
-        ...
-        $OBJ.QueryRowContext($CTX, $QUERY, ...)
+    - patterns:
+        - pattern-either:
+            - pattern: $QUERY = fmt.Fprintf($F, "$SQLSTR", ...)
+            - pattern: $QUERY = fmt.Sprintf("$SQLSTR", ...)
+            - pattern: $QUERY = fmt.Printf("$SQLSTR", ...)
+            - pattern: $QUERY = $X + ...
+        - pattern-either:
+            - pattern-inside: |
+                func $FUNC(...) {
+                    ...
+                    $OBJ.Query($QUERY, ...)
```

Comment:
This will perform much better - now that we're scoping to a function, we probably need a trailing ellipsis to make sure we don't miss e.g.

```go
func postgresQueryFmt(r *http.Request) {
    ctx := context.Background()
    customerId := r.URL.Query().Get("id")
	// ruleid: string-formatted-query
	query := fmt.Sprintf("SELECT number, expireDate, cvv FROM creditcards WHERE customerId = %s", customerId)

    row, _ := postgresDb.QueryRow(ctx, query)
    someotherResult := doSomethingElse()
}
```

Reply:
Makes sense to me. Added.